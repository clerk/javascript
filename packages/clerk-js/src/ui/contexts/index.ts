export * from './CoreClerkContext';
export * from './CoreUserContext';
export * from './CoreOrganizationContext';
export * from './ClerkUIComponentsContext';
export * from './EnvironmentContext';
export * from './OptionsContext';
export * from './CoreSessionContext';
export * from './CoreClientContext';
