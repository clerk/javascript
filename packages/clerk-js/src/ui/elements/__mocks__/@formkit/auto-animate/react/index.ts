const useAutoAnimate = () => [null];
export { useAutoAnimate };
