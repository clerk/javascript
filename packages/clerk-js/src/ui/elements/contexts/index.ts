export * from './CardStateContext';
export * from './FlowMetadataContext';
