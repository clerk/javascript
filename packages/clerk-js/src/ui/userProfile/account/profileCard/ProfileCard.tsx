import { List } from '@clerk/shared/components/list';
import { TitledCard } from '@clerk/shared/components/titledCard';
import type { UserResource } from '@clerk/types';
import { ExternalAccountResource } from '@clerk/types';
import React from 'react';
import { svgUrl } from 'ui/common/constants';
import { useCoreUser,useEnvironment } from 'ui/contexts';
import { useNavigate } from 'ui/hooks';

import { AvatarWithUploader } from '../avatarWithUploader';
import { ProfileEmailAddresses } from './ProfileEmailAddresses';
import { ProfilePhoneNumbers } from './ProfilePhoneNumbers';

// Since only Metamask wallet is supported for now, we assume that the user has only
// one Web3 wallet address. When we add support for more wallets, this logic should be
// refactored.
function getWeb3WalletAddress(user: UserResource): string {
  if (user.web3Wallets?.length > 0) {
    return user.web3Wallets[0] ? user.web3Wallets[0].web3Wallet : '';
  }
  return '';
}

export function ProfileCard(): JSX.Element {
  const { authConfig } = useEnvironment();
  const user = useCoreUser();
  const { navigate } = useNavigate();
  const web3Wallet = getWeb3WalletAddress(user);

  const buildWebWalletRow = () => (
    <List.Item
      className='cl-list-item'
      key='web-wallet-name'
      itemTitle='Wallet address'
      detail={false}
    >
      {web3Wallet}
    </List.Item>
  );

  const buildUsernameRow = () => (
    <List.Item
      className='cl-list-item'
      key='user-name'
      itemTitle='Username'
      onClick={() => navigate('username')}
    >
      {user.username}
    </List.Item>
  );

  const buildConnectedAccountsRow = () => (
    <List.Item
      className='cl-list-item'
      key='connected-accounts'
      itemTitle='Connected accounts'
      onClick={() => navigate('connected-accounts')}
    >
      {user.externalAccounts.length === 0 ? (
        <div className='cl-empty-list-item'>None</div>
      ) : (
        <div className='cl-list-item-entry'>
          {user.externalAccounts.map(
            (externalAccount: ExternalAccountResource) => (
              <div key={externalAccount.id}>
                <img
                  alt={externalAccount.providerTitle()}
                  src={svgUrl(externalAccount.provider)}
                  className='cl-left-icon-wrapper'
                />
                {externalAccount.emailAddress}
              </div>
            ),
          )}
        </div>
      )}
    </List.Item>
  );

  const avatarRow = (
    <List.Item
      className='cl-list-item'
      key='photo'
      itemTitle='Photo'
      hoverable={false}
      detail={false}
    >
      <AvatarWithUploader
        size={64}
        firstName={user.firstName}
        lastName={user.lastName}
        name={user.fullName}
        profileImageUrl={user.profileImageUrl}
        className='cl-image'
      />
    </List.Item>
  );

  const showWebWallet = !!web3Wallet;
  const showUsername = authConfig.username === 'on';
  const showEmail = authConfig.emailAddress === 'on';
  const showPhone = authConfig.phoneNumber === 'on';

  return (
    <TitledCard
      className='cl-themed-card'
      title='Profile'
      subtitle='Manage profile settings'
    >
      <List className='cl-titled-card-list'>
        {avatarRow}
        {showUsername && buildUsernameRow()}
        {showEmail && <ProfileEmailAddresses />}
        {showPhone && <ProfilePhoneNumbers />}
        {buildConnectedAccountsRow()}
        {showWebWallet && buildWebWalletRow()}
      </List>
    </TitledCard>
  );
}
