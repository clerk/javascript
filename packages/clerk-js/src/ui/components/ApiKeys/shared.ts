import { useClerk } from '@clerk/shared/react';
import { useState } from 'react';

import { useFetch } from '../../hooks';
import type { Expiration } from './CreateApiKeyForm';

function getTimeLeftInSeconds(expirationOption: Expiration) {
  if (expirationOption === 'never') {
    return;
  }
  const now = new Date();
  const future = new Date(now);
  if (expirationOption === '30d') {
    future.setDate(future.getDate() + 30);
  } else if (expirationOption === '90d') {
    future.setDate(future.getDate() + 90);
  } else {
    throw new Error('TODO: Improve time handling');
  }
  const diffInMs = future.getTime() - now.getTime();
  const diffInSecs = Math.floor(diffInMs / 1000);
  return diffInSecs;
}

export function useApiKeys({ subject, perPage = 5 }: { subject: string; perPage?: number }) {
  const clerk = useClerk();
  const {
    data: apiKeys,
    isLoading,
    revalidate,
  } = useFetch(clerk.getApiKeys, { subject }, undefined, `api-key-source-${subject}`);
  const [page, setPage] = useState(1);
  const [search, setSearch] = useState('');
  const itemsPerPage = perPage;

  const filteredApiKeys = (apiKeys ?? []).filter(key => key.name.toLowerCase().includes(search.toLowerCase()));
  const itemCount = filteredApiKeys.length;
  const pageCount = Math.max(1, Math.ceil(itemCount / itemsPerPage));
  const startingRow = itemCount > 0 ? (page - 1) * itemsPerPage + 1 : 0;
  const endingRow = Math.min(page * itemsPerPage, itemCount);
  const paginatedApiKeys = filteredApiKeys.slice(startingRow - 1, endingRow);

  const handleCreate = async (params: {
    name: string;
    description?: string;
    expiration: Expiration;
    closeFn: () => void;
  }) => {
    await clerk.createApiKey({
      name: params.name,
      creationReason: params.description,
      secondsUntilExpiration: getTimeLeftInSeconds(params.expiration),
    });
    params.closeFn();
    revalidate();
  };

  const revokeApiKey = async (apiKeyID: string) => {
    await clerk.revokeApiKey({ apiKeyID, revocationReason: 'Revoked by user' });
    setPage(1);
    revalidate();
  };

  return {
    apiKeys: paginatedApiKeys,
    isLoading: isLoading ?? false,
    revokeApiKey,
    search,
    setSearch,
    page,
    setPage,
    pageCount,
    itemCount,
    startingRow,
    endingRow,
    handleCreate,
  };
}
