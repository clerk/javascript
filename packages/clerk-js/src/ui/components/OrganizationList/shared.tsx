import type { UserOrganizationInvitationResource } from '@clerk/types';
import type { PropsWithChildren } from 'react';
import { forwardRef } from 'react';

import { Box, Button, Col, descriptors, Flex, Spinner } from '../../customizables';
import { OrganizationPreview, PreviewButton } from '../../elements';
import { ArrowRightIcon } from '../../icons';
import type { ThemableCssProp } from '../../styledSystem';
import { common } from '../../styledSystem';

export const PreviewListItems = (props: PropsWithChildren) => {
  return (
    <Col
      elementDescriptor={descriptors.organizationListPreviewItems}
      sx={t => ({
        maxHeight: `calc(8 * ${t.sizes.$12})`,
        overflowY: 'auto',
        ...common.unstyledScrollbar(t),
      })}
    >
      {props.children}
    </Col>
  );
};

const sharedStyles: ThemableCssProp = t => ({
  height: t.space.$14,
  padding: `${t.space.$2} ${t.space.$8}`,
});

export const PreviewListItem = (
  props: PropsWithChildren<{
    organizationData: UserOrganizationInvitationResource['publicOrganizationData'];
  }>,
) => {
  return (
    <Flex
      align='center'
      gap={2}
      sx={[
        _ => ({
          minHeight: 'unset',
          justifyContent: 'space-between',
        }),
        sharedStyles,
      ]}
      elementDescriptor={descriptors.organizationListPreviewItem}
    >
      <OrganizationPreview
        elementId='organizationList'
        avatarSx={t => ({ width: t.sizes.$10, height: t.sizes.$10 })}
        mainIdentifierSx={t => ({
          fontSize: t.fontSizes.$xl,
          color: t.colors.$colorText,
        })}
        organization={props.organizationData}
      />
      {props.children}
    </Flex>
  );
};

export const PreviewListSpinner = forwardRef<HTMLDivElement>((_, ref) => {
  return (
    <Box
      ref={ref}
      sx={t => ({
        width: '100%',
        height: t.space.$12,
        position: 'relative',
      })}
    >
      <Box
        sx={{
          margin: 'auto',
          position: 'absolute',
          left: '50%',
          top: '50%',
          transform: 'translateY(-50%) translateX(-50%)',
        }}
      >
        <Spinner
          size='md'
          colorScheme='primary'
        />
      </Box>
    </Box>
  );
});

export const PreviewListItemButton = (props: Parameters<typeof Button>[0]) => {
  return (
    <Button
      elementDescriptor={descriptors.organizationListPreviewItemActionButton}
      textVariant='buttonExtraSmallBold'
      variant='outline'
      colorScheme='neutral'
      size='sm'
      {...props}
    />
  );
};

export const OrganizationListPreviewButton = (props: PropsWithChildren<{ onClick: () => void | Promise<void> }>) => {
  return (
    <PreviewButton
      elementDescriptor={descriptors.organizationListPreviewButton}
      sx={[sharedStyles]}
      icon={ArrowRightIcon}
      iconProps={{
        size: 'lg',
      }}
      showIconOnHover={false}
      {...props}
    />
  );
};
