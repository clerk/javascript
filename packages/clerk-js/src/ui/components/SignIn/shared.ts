import { isClerkRuntimeError, isUserLockedError } from '@clerk/shared/error';
import { useClerk } from '@clerk/shared/react';
import { useCallback, useEffect } from 'react';

import { clerkInvalidFAPIResponse } from '../../../core/errors';
import { __internal_WebAuthnAbortService } from '../../../utils/passkeys';
import { useCoreSignIn, useSignInContext } from '../../contexts';
import { useCardState } from '../../elements';
import { useSupportEmail } from '../../hooks/useSupportEmail';
import { handleError } from '../../utils';

function useHandleAuthenticateWithPasskey(onSecondFactor: () => Promise<unknown>) {
  const card = useCardState();
  // @ts-expect-error -- private method for the time being
  const { setActive, __internal_navigateWithError } = useClerk();
  const supportEmail = useSupportEmail();
  const { afterSignInUrl } = useSignInContext();
  const { authenticateWithPasskey } = useCoreSignIn();

  useEffect(() => {
    return () => {
      __internal_WebAuthnAbortService.abort();
    };
  }, []);

  return useCallback(async (...args: Parameters<typeof authenticateWithPasskey>) => {
    try {
      const res = await authenticateWithPasskey(...args);
      switch (res.status) {
        case 'complete':
          return setActive({ session: res.createdSessionId, redirectUrl: afterSignInUrl });
        case 'needs_second_factor':
          return onSecondFactor();
        default:
          return console.error(clerkInvalidFAPIResponse(res.status, supportEmail));
      }
    } catch (err) {
      const { flow } = args[0] || {};

      if (isClerkRuntimeError(err)) {
        // In any case if the call gets aborted we should skip showing an error. This prevents updating the state of unmounted components.
        if (err.code === 'passkey_operation_aborted') {
          return;
        }
        // In case of autofill, if retrieval of credentials is cancelled by the user avoid showing errors as it results to pour UX.
        if (flow === 'autofill' && err.code === 'passkey_retrieval_cancelled') {
          return;
        }
      }

      if (isUserLockedError(err)) {
        return __internal_navigateWithError('..', err.errors[0]);
      }
      handleError(err, [], card.setError);
    }
  }, []);
}

export { useHandleAuthenticateWithPasskey };
