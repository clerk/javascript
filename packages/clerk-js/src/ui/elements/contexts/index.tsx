import { createContextAndHook } from '@clerk/shared/react';
import type { ClerkAPIError, ClerkRuntimeError } from '@clerk/types';
import { FloatingTree, useFloatingParentNodeId } from '@floating-ui/react';
import React from 'react';

import { useLocalizations } from '../../customizables';
import { useSafeState } from '../../hooks';

type Status = 'idle' | 'loading' | 'error';
type Metadata = string | undefined;
type State = { status: Status; metadata: Metadata; error: string | undefined };
type CardStateCtxValue = {
  state: State;
  setState: React.Dispatch<React.SetStateAction<State>>;
};

const [CardStateCtx, _useCardState] = createContextAndHook<CardStateCtxValue>('CardState');

export const CardStateProvider = (props: React.PropsWithChildren<any>) => {
  const { translateError } = useLocalizations();

  const [state, setState] = useSafeState<State>({
    status: 'idle',
    metadata: undefined,
    error: translateError(window?.Clerk?.__internal_last_error || undefined),
  });

  const value = React.useMemo(() => ({ value: { state, setState } }), [state, setState]);
  return <CardStateCtx.Provider value={value}>{props.children}</CardStateCtx.Provider>;
};

export const useCardState = () => {
  const { state, setState } = _useCardState();
  const { translateError } = useLocalizations();

  const setIdle = (metadata?: Metadata) => setState(s => ({ ...s, status: 'idle', metadata }));
  const setError = (metadata: ClerkRuntimeError | ClerkAPIError | Metadata | string) =>
    setState(s => ({ ...s, error: translateError(metadata) }));
  const setLoading = (metadata?: Metadata) => setState(s => ({ ...s, status: 'loading', metadata }));
  const runAsync = async <T = unknown,>(cb: Promise<T> | (() => Promise<T>), metadata?: Metadata) => {
    setLoading(metadata);
    return (typeof cb === 'function' ? cb() : cb)
      .then(res => {
        return res;
      })
      .finally(() => setIdle(metadata));
  };

  return {
    setIdle,
    setError,
    setLoading,
    runAsync,
    loadingMetadata: state.status === 'loading' ? state.metadata : undefined,
    error: state.error ? state.error : undefined,
    isLoading: state.status === 'loading',
    isIdle: state.status === 'idle',
    state,
  };
};

export const withCardStateProvider = <T,>(Component: React.ComponentType<T>) => {
  return (props: T) => {
    return (
      <CardStateProvider>
        {/* @ts-expect-error */}
        <Component {...props} />
      </CardStateProvider>
    );
  };
};

export type FlowMetadata = {
  flow:
    | 'signIn'
    | 'signUp'
    | 'userButton'
    | 'userProfile'
    | 'userVerification'
    | 'organizationProfile'
    | 'createOrganization'
    | 'organizationSwitcher'
    | 'organizationList'
    | 'oneTap'
    | 'blankCaptcha'
    | 'waitlist'
    | 'checkout'
    | 'planDetails'
    | 'pricingTable';
  part?:
    | 'start'
    | 'emailCode'
    | 'phoneCode'
    | 'phoneCode2Fa'
    | 'totp2Fa'
    | 'backupCode2Fa'
    | 'password'
    | 'resetPassword'
    | 'emailLink'
    | 'emailLinkVerify'
    | 'emailLinkStatus'
    | 'alternativeMethods'
    | 'forgotPasswordMethods'
    | 'passwordPwnedMethods'
    | 'havingTrouble'
    | 'ssoCallback'
    | 'popupCallback'
    | 'popover'
    | 'complete'
    | 'accountSwitcher';
};

const [FlowMetadataCtx, useFlowMetadata] = createContextAndHook<FlowMetadata>('FlowMetadata');

export const FlowMetadataProvider = (props: React.PropsWithChildren<FlowMetadata>) => {
  const { flow, part } = props;
  const value = React.useMemo(() => ({ value: props }), [flow, part]);
  return <FlowMetadataCtx.Provider value={value}>{props.children}</FlowMetadataCtx.Provider>;
};

export const withFloatingTree = <T,>(Component: React.ComponentType<T>): React.ComponentType<T> => {
  return (props: T) => {
    const parentId = useFloatingParentNodeId();
    if (parentId == null) {
      return (
        <FloatingTree>
          {/* @ts-expect-error */}
          <Component {...props} />
        </FloatingTree>
      );
    }

    /* @ts-expect-error */
    return <Component {...props} />;
  };
};

export { useFlowMetadata };
