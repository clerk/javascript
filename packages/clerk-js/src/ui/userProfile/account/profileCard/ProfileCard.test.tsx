import { renderJSON } from '@clerk/shared/testUtils';
import { EmailAddressResource, UserResource } from '@clerk/types';
import * as React from 'react';
import { PartialDeep } from 'type-fest';
import { useEnvironment } from 'ui/contexts';

import { ProfileCard } from './ProfileCard';

jest.mock('ui/hooks', () => {
  return {
    useNavigate: () => ({ navigate: jest.fn() }),
  };
});

jest.mock('ui/contexts', () => {
  return {
    useEnvironment: jest.fn(),
    useCoreUser: (): PartialDeep<UserResource> => {
      return {
        isPrimaryIdentification: jest.fn(() => true),
        emailAddresses: [
          {
            id: 1,
            emailAddress: 'clerk@clerk.dev',
            verification: { status: 'verified' },
          } as any as EmailAddressResource,
          {
            id: 2,
            emailAddress: 'clerk-unverified@clerk.dev',
            verification: { status: 'unverified' },
          } as any as EmailAddressResource,
        ],
        phoneNumbers: [],
        externalAccounts: [],
        web3Wallets: [{ web3Wallet: '0x123456789' }],
      };
    },
    useCoreClerk: jest.fn(),
  };
});

(useEnvironment as jest.Mock).mockImplementation(() => ({
  displayConfig: {},
  userSettings: {
    attributes: {
      email_address: {
        enabled: true,
      },
      phone_number: {
        enabled: true,
      },
      username: {
        enabled: false,
      },
      web3_wallet: {
        enabled: false,
      },
    },
  },
}));

describe('<ProfileCard/>', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders the ProfileCard', () => {
    const tree = renderJSON(<ProfileCard />);
    expect(tree).toMatchSnapshot();
  });

  it('renders the ProfileCard ', () => {
    (useEnvironment as jest.Mock).mockImplementation(() => ({
      displayConfig: {},
      userSettings: {
        attributes: {
          email_address: {
            enabled: true,
          },
          phone_number: {
            enabled: true,
          },
          username: {
            enabled: false,
          },
          web3_wallet: {
            enabled: true,
          },
        },
      },
    }));

    const tree = renderJSON(<ProfileCard />);
    expect(tree).toMatchSnapshot();
  });
});
