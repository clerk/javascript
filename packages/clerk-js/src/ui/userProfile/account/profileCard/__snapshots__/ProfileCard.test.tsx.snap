// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProfileCard/> renders the ProfileCard  1`] = `
<div
  className="card card cl-themed-card"
>
  <div
    className="titleContainer"
  >
    <h1
      className="title"
    >
      Profile
    </h1>
    <p
      className="subtitle"
    >
      Manage profile settings
    </p>
  </div>
  <div
    className="cl-titled-card-list"
  >
    <div
      className="cl-list-item elementContainer"
      disabled={false}
    >
      <div
        className="title"
      >
        Photo
      </div>
      <div
        className="listItem line"
      >
        <div
          className="start"
        >
          <div
            className="uploaderWrapper wrapper"
            onClick={[Function]}
          >
            <img
              alt=""
              className="avatar cl-image"
              height={64}
              onError={[Function]}
              src="https://www.gravatar.com/avatar?d=mp"
              title=""
              width={64}
            />
            <div
              className="touchContainer"
            >
              <svg
                data-filename="../shared/assets/icons/camera.svg"
              />
            </div>
            <div
              className="hoverContainer"
            >
              <svg
                data-filename="../shared/assets/icons/camera.svg"
              />
            </div>
          </div>
        </div>
        <div
          className="iconContainer"
        />
      </div>
    </div>
    <button
      className="cl-list-item elementContainer button hoverable"
      disabled={false}
      onClick={[Function]}
    >
      <div
        className="title"
      >
        Email
      </div>
      <div
        className="listItem line"
      >
        <div
          className="start"
        >
          <div
            className="cl-list-item-entry"
          >
            <div>
              clerk@clerk.dev
              <span
                className="tag primary sm"
              >
                Primary
              </span>
              <span
                className="tag success sm cl-tag"
              >
                Verified
              </span>
            </div>
            <div>
              clerk-unverified@clerk.dev
              <span
                className="tag primary sm"
              >
                Primary
              </span>
              <span
                className="tag error sm cl-tag"
              >
                Unverified
              </span>
            </div>
          </div>
        </div>
        <div
          className="iconContainer"
        >
          <svg
            data-filename="../shared/assets/icons/arrow-right.svg"
          />
        </div>
      </div>
    </button>
    <button
      className="cl-list-item elementContainer button hoverable"
      disabled={false}
      onClick={[Function]}
    >
      <div
        className="title"
      >
        Phone number
      </div>
      <div
        className="listItem line"
      >
        <div
          className="start"
        >
          <div
            className="cl-empty-list-item"
          >
            None
          </div>
        </div>
        <div
          className="iconContainer"
        >
          <svg
            data-filename="../shared/assets/icons/arrow-right.svg"
          />
        </div>
      </div>
    </button>
    <button
      className="cl-list-item elementContainer button hoverable"
      disabled={false}
      onClick={[Function]}
    >
      <div
        className="title"
      >
        Connected accounts
      </div>
      <div
        className="listItem line"
      >
        <div
          className="start"
        >
          <div
            className="cl-empty-list-item"
          >
            None
          </div>
        </div>
        <div
          className="iconContainer"
        >
          <svg
            data-filename="../shared/assets/icons/arrow-right.svg"
          />
        </div>
      </div>
    </button>
    <button
      className="cl-list-item elementContainer button hoverable"
      disabled={false}
    >
      <div
        className="title"
      >
        Wallet address
      </div>
      <div
        className="listItem line"
      >
        <div
          className="start"
        >
          0x123456789
        </div>
        <div
          className="iconContainer"
        />
      </div>
    </button>
  </div>
</div>
`;

exports[`<ProfileCard/> renders the ProfileCard 1`] = `
<div
  className="card card cl-themed-card"
>
  <div
    className="titleContainer"
  >
    <h1
      className="title"
    >
      Profile
    </h1>
    <p
      className="subtitle"
    >
      Manage profile settings
    </p>
  </div>
  <div
    className="cl-titled-card-list"
  >
    <div
      className="cl-list-item elementContainer"
      disabled={false}
    >
      <div
        className="title"
      >
        Photo
      </div>
      <div
        className="listItem line"
      >
        <div
          className="start"
        >
          <div
            className="uploaderWrapper wrapper"
            onClick={[Function]}
          >
            <img
              alt=""
              className="avatar cl-image"
              height={64}
              onError={[Function]}
              src="https://www.gravatar.com/avatar?d=mp"
              title=""
              width={64}
            />
            <div
              className="touchContainer"
            >
              <svg
                data-filename="../shared/assets/icons/camera.svg"
              />
            </div>
            <div
              className="hoverContainer"
            >
              <svg
                data-filename="../shared/assets/icons/camera.svg"
              />
            </div>
          </div>
        </div>
        <div
          className="iconContainer"
        />
      </div>
    </div>
    <button
      className="cl-list-item elementContainer button hoverable"
      disabled={false}
      onClick={[Function]}
    >
      <div
        className="title"
      >
        Email
      </div>
      <div
        className="listItem line"
      >
        <div
          className="start"
        >
          <div
            className="cl-list-item-entry"
          >
            <div>
              clerk@clerk.dev
              <span
                className="tag primary sm"
              >
                Primary
              </span>
              <span
                className="tag success sm cl-tag"
              >
                Verified
              </span>
            </div>
            <div>
              clerk-unverified@clerk.dev
              <span
                className="tag primary sm"
              >
                Primary
              </span>
              <span
                className="tag error sm cl-tag"
              >
                Unverified
              </span>
            </div>
          </div>
        </div>
        <div
          className="iconContainer"
        >
          <svg
            data-filename="../shared/assets/icons/arrow-right.svg"
          />
        </div>
      </div>
    </button>
    <button
      className="cl-list-item elementContainer button hoverable"
      disabled={false}
      onClick={[Function]}
    >
      <div
        className="title"
      >
        Phone number
      </div>
      <div
        className="listItem line"
      >
        <div
          className="start"
        >
          <div
            className="cl-empty-list-item"
          >
            None
          </div>
        </div>
        <div
          className="iconContainer"
        >
          <svg
            data-filename="../shared/assets/icons/arrow-right.svg"
          />
        </div>
      </div>
    </button>
    <button
      className="cl-list-item elementContainer button hoverable"
      disabled={false}
      onClick={[Function]}
    >
      <div
        className="title"
      >
        Connected accounts
      </div>
      <div
        className="listItem line"
      >
        <div
          className="start"
        >
          <div
            className="cl-empty-list-item"
          >
            None
          </div>
        </div>
        <div
          className="iconContainer"
        >
          <svg
            data-filename="../shared/assets/icons/arrow-right.svg"
          />
        </div>
      </div>
    </button>
  </div>
</div>
`;
